[FadhilPM] iP

FadhilPM opened the pull request:
# Bezdelnik: A Task Manager
> 

Bezdelnik is a _flexible_ ~~OOP dumpster-fire~~ **simple** task management application designed to help you organize your life.

## Why Choose Bezdelnik?

* **3 Simple Task Types** - Create todos, deadlines, and events
* **Dual Interface** - Use either GUI or CLI
* **Immutable Data Management** - Data is always safe and consistent
* **Automatic Saving** - Never lose tasks again

### Getting Started in 3 Easy Steps

1. Download the latest release from [the GitHub repo](https://github.com/FadhilPM/bezdelnik)
2. Run the application using `java -jar bezdelnik.jar`
3. Start adding tasks with simple commands 😉
 
## Feature Roadmap

- [x] Basic task management
- [x] Date and time handling
- [x] Task filtering and sorting
- [x] Recurring tasks (someday, maybe?)

Try Bezdelnik today!

## Sample code 
Here's Bezdelnik's `Main` method.

```java
public class Main extends Application {

    private Bezdelnik bezdelnik = new Bezdelnik();

    @Override
    public void start(Stage stage) {
        try {
            FXMLLoader fxmlLoader = new FXMLLoader(Main.class.getResource("/view/MainWindow.fxml"));
            AnchorPane ap = fxmlLoader.load();
            Scene scene = new Scene(ap);
            stage.setScene(scene);
            fxmlLoader.<MainWindow>getController().setBezdelnik(bezdelnik);
            stage.show();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

FadhilPM committed:
Initial commit: Renamed bot and added skeletal methods.

FadhilPM committed:
Added echoing of user input

FadhilPM committed:
Added storage of text by user display of stored text on demand.

FadhilPM committed:
Level-3

Added ability to mark and unmark tasks as done.
Used regex to make command interpretation slightly more flexible.
Rudimentary progress towards Level-5 with basic exception handling for common mistakes.

FadhilPM committed:
Level-4 inprogress

Level 4 changes fulfilled, but code is very rough around the edges and needs to be refactored.
Task is now abstract, added support for 3 different types of Tasks (Todos, Deadlines, Events)

FadhilPM committed:
ctrl+d exception handling

used a try-catch to interpret ctrl+d as an exit

FadhilPM committed:
Level-6

deletion of tasks added.

FadhilPM committed:
more OOP

Task list handling moved to new Taskman class

FadhilPM committed:
Minor cleanup

FadhilPM committed:
Partial progress

made stringParser method pure

FadhilPM committed:
Exception cleanup

removed redundant try-catch blocks to reduce length and improve readability of code.

FadhilPM committed:
Code cleanup

Streams used for main input loop. Code now clean enough to claim compliance with levels 4 and 5.

FadhilPM committed:
Code cleanup

Streams used for main input loop. Code cleaned up to claim compliance with levels 4 and 5.

FadhilPM committed:
Merge branch 'Purification'

FadhilPM committed:
Refactor

Moved string parser to separate class.

FadhilPM committed:
Merge tag 'Level-4'

FadhilPM committed:
Refactor

Laying groundwork for Level-7 and Level-9
Task now implements serializable and has the ability to return the command that created it.

FadhilPM committed:
Added ability to check if task is done

FadhilPM committed:
Improved style compliance

Style compliance improved in accordance with Google's Java Style Guide.

FadhilPM committed:
Markdown expts

FadhilPM committed:
Automated text ui testing

Finally added automated text ui testing!

Scripts have been modified to ensure the use of JDK 17 even if a newer JDK is being used by default.

FadhilPM committed:
Compliance with CS2103/SE-EDU coding standard

FadhilPM committed:
Add equals() method to Taskman

FadhilPM committed:
Add code to convert task list to plaintext commands

FadhilPM committed:
Rename CommandParser to Parser in line with A-MoreOOP

FadhilPM committed:
Fully implement save/restore functionality

FadhilPM committed:
Checkstyle compliance

FadhilPM committed:
Merge branch 'branch-Level-7'

FadhilPM committed:
Implement Level-9: Find

FadhilPM committed:
Merge branch 'branch-Level-9'

FadhilPM committed:
Implement rudimentary datetime handling

FadhilPM committed:
Merge branch 'branch-Level-8'

FadhilPM committed:
Update .gitignore

FadhilPM committed:
Merge branch 'add-gradle-support'

FadhilPM committed:
Move source code to package

FadhilPM committed:
Merge branch 'A-Packages'

FadhilPM committed:
Add preliminary Gradle support

FadhilPM committed:
Create UI utility class

FadhilPM committed:
Merge branch 'Functional'

CJ-783 commented on src/main/java/Bezdelnik/Task.java:
```diff
@@ -0,0 +1,32 @@
+        return isDone;
+    }
+
+    public boolean contains(String s) {
```
Perhaps the name of the method could be more informative?

CJ-783 commented on src/main/java/Bezdelnik/Taskman.java:
```diff
@@ -0,0 +1,87 @@
+        this.taskList = taskStream.toList();
+    }
+
+    public Taskman add(Task t) {
```
Perhaps the name of the method could be more informative? I noticed the same issue in several other places too

CJ-783 commented on src/main/java/Bezdelnik/Deadline.java:
```diff
@@ -0,0 +1,44 @@
+        this(description, false, by);
+    }
+
+    Deadline(String description, boolean isDone, String byString) {
```
I like how the method name sounds like booleans

CJ-783 commented on src/main/java/Bezdelnik/Task.java:
```diff
@@ -0,0 +1,32 @@
+
+abstract class Task implements Serializable {
+    protected final String description;
+    protected final boolean isDone;
```
I like how the variable name sounds like booleans

CJ-783 reviewed (commented):
Overall, LGTM! 🔥 🔥 🔥 

Just a few nits to fix

FadhilPM committed:
Improve checkstyle compliance

FadhilPM committed:
Add javadoc documentation

FadhilPM committed:
Merge branch 'A-JavaDoc'

FadhilPM committed:
Comply with SE-EDU coding standard

FadhilPM committed:
Patch potential memory leak

FadhilPM committed:
Merge branch 'A-CodingStandard'

FadhilPM committed:
Fix unused imports

FadhilPM committed:
Add rudimentary JUnit tests

FadhilPM committed:
Update gitignore

FadhilPM committed:
Explicitly declare classes public

GaaneshT commented on src/main/java/bezdelnik/Bezdelnik.java:
```diff
@@ -0,0 +1,55 @@
+     *
+     * @param sessionStatus Initial session status message.
+     */
+    private static void inputLoop(String sessionStatus) {
+        Ui.greet(sessionStatus);
+        Scanner sc = new Scanner(System.in);
+        sc.useDelimiter("\n")
+            .tokens()
+            .map(input -> input.strip())
+            .takeWhile(input -> !input.matches("(bye|(/)?ex(it)?)"))
+            .map(input -> Parser.parse(input, taskman))
+            .forEach(pair -> {
+                Ui.print(pair.first());
+                taskman = pair.second();
+                try {
+                    Storage.writeTaskmanToFile(taskman, saveLocation);
+                } catch (Throwable e) {
+                    Ui.print(String.format("Unknown exception when saving data.", e.toString()));
+                }
+            }
+            );
+        sc.close();
+        Ui.bye();
+    }
+}
```
I like the use of streams here! It's good approach :)

GaaneshT commented on src/main/java/bezdelnik/Parser.java:
```diff
@@ -0,0 +1,185 @@
+     * @return A Pair with a confirmation message and the updated task manager.
+     */
+    private static Pair<String, Taskman> handleTodo(String input, Taskman taskman) {
+        String todoInput = input.substring(5);
```
I think you might want to avoid using substring(n) to extract arguments. It expects fix length commands which would break if the user adds an extra space. I think using regex or split would be better here ! :D

GaaneshT commented on src/main/java/bezdelnik/Parser.java:
```diff
@@ -0,0 +1,185 @@
+                return handleMark(input, taskman);
+            case "u":
+            case "unmark":
+                return handleUnmark(input, taskman);
+            case "rem":
+            case "remove":
+            case "del":
+            case "delete":
+                return handleRemove(input, taskman);
```
I like this! Makes it more robust :)

GaaneshT commented on src/main/java/bezdelnik/Taskman.java:
```diff
@@ -0,0 +1,183 @@
+     * @return The task at the specified index.
+     * @throws BezdelnikException If the index is out of bounds.
+     */
+    public Task get(int i) throws BezdelnikException {
+        try {
+            return taskList.get(i);
+        } catch (Throwable t) {
+            throw new BezdelnikException(t.toString());
+        }
+    }
```
I think the method name could be improved here! Perhaps getIndex() ?

GaaneshT commented on src/main/java/bezdelnik/Taskman.java:
```diff
@@ -0,0 +1,183 @@
+     * @param pt The predicate to filter tasks.
+     * @return A new Taskman instance with filtered tasks.
+     */
+    public Taskman filter(Predicate<Task> pt) {
```
Same thing here, could perhaps change to filterList()? Makes it more readable!

GaaneshT commented on src/main/java/bezdelnik/Event.java:
Clean code here, good job!

GaaneshT commented on src/main/java/bezdelnik/Bezdelnik.java:
I like your use of streams here!

GaaneshT reviewed (commented):
LGTM! I really like your use of streams, it's a unique approach! 
Good job :D

FadhilPM committed:
Improve robustness of Parser

FadhilPM committed:
Add rudimentary task sorting capability

FadhilPM committed:
Add archiving capability

FadhilPM committed:
Refactor in preparation for GUI addition

FadhilPM committed:
Add rudimentary GUI

FadhilPM committed:
Merge branch 'branch-Level-10'

FadhilPM committed:
Add load success/failure message to GUI

FadhilPM committed:
Add basic asserts

FadhilPM committed:
Add checkstyle config files

FadhilPM committed:
Add additional asserts

FadhilPM committed:
Merge branch 'A-Assertions'

FadhilPM committed:
Taskman: replace ternary operators with if/else statements

The Taskman class uses several ternary operators that make the code harder
to read and understand.

Let's convert all ternary operators to if/else blocks for the sake of readability.

FadhilPM committed:
Storage: restructure code logically

The Storage class clumps code into subsequent lines, which makes the purpose of said code more difficult discern.

Let's include line spacing to improve the readability of Storage.

FadhilPM committed:
Parser: Improve code readability

The Parser class currently has dense code, which may be difficult to read and understand.

Let's add line breaks at logical locations to improve readability.

FadhilPM committed:
Task: make Task no longer implement Serializable

The Task abstract class currently implements Serializable, which was intended as an easy means of archiving tasks to file.

Let's remove the implementation of the Serializable interface as this functionality has remained unused and is dead code.

FadhilPM committed:
Add further assertions

FadhilPM committed:
Merge pull request #2 from FadhilPM/A-Assertions

Add further assertions

FadhilPM committed:
Merge branch 'master' into A-CodeQuality

FadhilPM committed:
Merge pull request #3 from FadhilPM/A-CodeQuality

Improve code quality

FadhilPM committed:
fix minor bug

FadhilPM committed:
Update build.gradle to package UI into fat JAR

FadhilPM committed:
Taskman: add ability to sort tasks by Comparator<? super Task>

FadhilPM committed:
Parser: add rudimentary error handling for Task creation

FadhilPM committed:
Ui: rename to ConsoleUi for additional clarity

FadhilPM committed:
Task: remove unused imports

FadhilPM committed:
Launcher: add command line option to start bot with CLI

FadhilPM committed:
MainWindow: add ability to exit GUI by text command

FadhilPM committed:
Bugfix: prevent GUI from opening once CLI chatbot closed

FadhilPM committed:
MainWindow: prevent user from sending empty input

FadhilPM committed:
Improve OOP adherence

FadhilPM committed:
Improve QoL

FadhilPM committed:
Merge branch 'OOPFix'

FadhilPM committed:
Move source files into logical packages

FadhilPM committed:
Move UI related code into ui package

FadhilPM committed:
Update JUnit tests to follow new packaging scheme

FadhilPM committed:
Merge branch 'FurtherPackaging'

FadhilPM committed:
Improve error-handling to be more useful to end user

FadhilPM committed:
refactor: Remove Storage class implementation

FadhilPM committed:
Storage: Break cyclic dependency on Parser

FadhilPM committed:
Remove unused Storage class

FadhilPM committed:
Parser: move all command logic to separate classes

FadhilPM committed:
Command: remove redundant catch blocks

FadhilPM committed:
Parser: refactor to return Commands

FadhilPM committed:
ReadStorage: Move all input parsing logic to Bezdelnik

FadhilPM committed:
Update JUnit tests

FadhilPM committed:
Fix bug when writing task list to file

FadhilPM committed:
Merge branch 'MVCFix'

FadhilPM committed:
Fix minor bug

FadhilPM committed:
Declare AI usage

FadhilPM committed:
Purge obsolete documentation

FadhilPM committed:
Add user guide

FadhilPM committed:
Update user interface

FadhilPM committed:
Update user guide

FadhilPM committed:
Update javadoc

FadhilPM committed:
Fix task sorting bug

FadhilPM committed:
Update user guide
